### PR TITLE
Prevent heap-use-after-free in Logger shutdown

### PR DESCRIPTION
**Pull request by danielhlarkin**
### Scope & Purpose

Prevents a heap-use-after-free in the Logger shutdown.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)

#### Related Information

- [x] There is an internal planning ticket: AR-60

### Testing & Verification

Jenkins: http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10867/

ASAN Runs:
http://jenkins.arangodb.biz:8080/job/arangodb-ANY-linux-asan/316/
http://jenkins.arangodb.biz:8080/job/arangodb-ANY-linux-asan/317/
http://jenkins.arangodb.biz:8080/job/arangodb-ANY-linux-asan/318/
http://jenkins.arangodb.biz:8080/job/arangodb-ANY-linux-asan/319/
http://jenkins.arangodb.biz:8080/job/arangodb-ANY-linux-asan/320/
http://jenkins.arangodb.biz:8080/job/arangodb-ANY-linux-asan/321/

### CODE REVIEW

**Comment by danielhlarkin** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10868/

File: lib/Logger/LogAppenderFile.cpp [hunk 1]
```diff
@@ -170,6 +171,8 @@ LogAppenderFile::LogAppenderFile(std::string const& filename, std::string const&
   _useColors = ((isatty(_fd) == 1) && Logger::getUseColor());
 }
 
+LogAppenderFile::~LogAppenderFile() { clear(); }
```
Comment (jsteemann) [inline lib/Logger/LogAppenderFile.cpp:174]: I am a bit worried about this. We can have multiple `LogAppenderFile` objects, for example, if we log to different output files (which is allowed and up to the user). Shutting down one of the `LogAppenderFile` objects would then close all of them, which probably isn't desired.

File: lib/Logger/LogAppenderFile.cpp [hunk 2]
```diff
@@ -269,19 +272,20 @@ void LogAppenderFile::reopenAll() {
...
-    // set the fd to "disabled"
-    std::get<0>(it) = -1;
-    // and also tell the appender of the file descriptor change
-    std::get<2>(it)->updateFd(-1);
-
     if (fd > STDERR_FILENO) {
+      // set the fd to "disabled"
+      std::get<0>(it) = -1;
+      // and also tell the appender of the file descriptor change
+      std::get<2>(it)->updateFd(-1);
```
Comment (jsteemann) [inline lib/Logger/LogAppenderFile.cpp:279]: I don't know if this change is ok...
Let's assume the server is a daemon and closed fds 0 and 1 on start.
Then fds 0 and 1 could be reused for any other file, even for custom logfiles. I think we need to do the `updateFd` call unconditionally of the file descriptor number.
I fear the file handling in the LogAppender* needs some more cleanup.

Comment (jsteemann) [reply lib/Logger/LogAppenderFile.cpp:279]: I will try to figure out what it currently does and then try to untangle the LogAppender* objects from the somehow managing the global list of appenders.

**Pull request closed without merging**